# RNAIndel

RNAIndel calls coding indels and classifies them into 
somatic, germline, and artifact from tumor RNA-Seq data.
Users can also use RNAIndel as a postprocessor to
classify indels called by their own callers. 
RNAIndel supports GRCh38 as well as GRCh37. 

## Prerequisites
Please make sure that the dependencies are satisfied before installing RNAIndel.
* [python>=3.5.2](https://www.python.org/downloads/)
    * [pandas>=0.23.0](https://pandas.pydata.org/) 
    * [scikit-learn>=0.18.1](http://scikit-learn.org/stable/install.html#)
    * [pysam>=0.13.0](https://pysam.readthedocs.io/en/latest/index.html)
* [java>=1.8.0](https://www.java.com/en/download/) 

## Installation
Install RNAIndel.
```
pip install rnaindel
```

## Data directory set up
Download datafile: [data_dir_37.tar.gz](http://ftp.stjude.org/pub/software/RNAIndel/data_dir_37.tar.gz) 
for GRCh37 and [data_dir_38.tar.gz](http://ftp.stjude.org/pub/software/RNAIndel/data_dir_38.tar.gz) for GRCh38.
Place the gzipped file under a directory of your choice and unpack it. 
```
tar xzvf data_dir_37.tar.gz  # for GRCh37
tar xzvf data_dir_38.tar.gz  # for GRCh38
```

## Usage ([demo](./sample_data)) 
RNAIndel has 3 modes: analysis, feature, training.

### Discover somatic indels from tumor RNA-Seq data
#### Built-in caller
RNAIndel calls indels  by the built-in caller [Bambino](https://academic.oup.com/bioinformatics/article/27/6/865/236751), which is optimized 
for RNA-Seq indel calling, and classifies into somatic, germline, and artifact. 
```
rnaindel analysis -b BAM -o OUTPUT_VCF -f FASTA -d DATA_DIR [other options]
```
#### Your caller
RNAIndel classifies indel entries in a VCF file generated by your caller (indel calling by the built-in caller will not be performed).
Specify the input VCF file by -v. <br> 
```
rnaindel analysis -b BAM -v INPUT_VCF -o OUTPUT_VCF -f FASTA -d DATA_DIR [other options]
```
#### Options
* ```-b``` input [STAR](https://academic.oup.com/bioinformatics/article/29/1/15/272537)-mapped BAM file (required)
* ```-v``` VCF file from other caller (required for using other callers, e.g., [GATK](https://software.broadinstitute.org/gatk/))
* ```-o``` output VCF file (required)
* ```-f``` reference genome (GRCh37 or 38) FASTA file (required)
* ```-d``` data directory contains trained models and databases (required) [Data directory set up](#data-directory-set-up) 
* ```-q``` STAR mapping quality MAPQ for unique mappers (default=255)
* ```-p``` number of cores (default=1)
* ```-m``` maximum heap space (default 6000m)
* ```-n``` user-defined panel of non-somatic indels in VCF format
* ```-l``` direcotry to store log files 
* ```-h``` print usage message
* ```--version``` print version


## Input BAM file
Please prepare your BAM file as follows:<br>

1. Map your reads with the STAR 2-pass mode to GRCh37 or 38.<br>
2. Add read groups, sort, mark duplicates, and index the BAM file with Picard.<br>

Please input the BAM file from Step 2 without caller-specific preprocessing such as indel realignment.<br>
Additional processing steps may prevent desired behavior.

## Panel of non-somatic indels (PONS)
Somatic prediction can be refined by applying a user-defined indel panel. Putative somatic indels found in the panel will be 
reclassified to germline or artifact, whichever has the higher probability. Indels predicted germline or artifact are not
subject to reclassification by PONS. Such panels can be compiled:
### from normal RNA-Seq data
RNA-Seq data may be a (ideally matched) single or a pooled dataset.<br>
1. Perform variant calling on the RNA-Seq data and generate a VCF file.<br>
2. Index the VCF with Tabix. <br>
### from a cohort dataset of tumor RNA-Seq and tumor/normal-paired DNA-Seq
In this approah, non-somatic indels recurrently misclassified as somatic are collected using a large cohort.<br>
1. Apply RNAIndel on the RNA-Seq data. <br>
2. Validate indels predicted as somatic (putative somatic indels) with the DNA-Seq data. <br>
3. Collect putative somatic indels which are validated as germline or artifact in N samples or more (recurrent non-somatic indels). <br>
4. Format the recurrent non-somatic indels in a VCF file and index with Tabix.<br>

A sample panel by the second approach is included in the [data package](#data-directory-set-up), which is compiled from a 
a cohort of 330 samples with RNA-Seq and T/N-paired WES & PCR-free WGS. When no custom panel is available, apply this panel by appending the following [option](#options):<br>
```
-n path/to/data_dir/non_somatic/non_somatic.vcf.gz
```

## Reference
1. Hagiwara, K., Ding, L., Edmonson, M.N., Rice, S.V., Newman, S., Meshinchi, S., Ries, R.E., Rusch, M., Zhang, J. 
RNAIndel: a machine-learning framework for discovery of somatic coding indels using tumor RNA-Seq data.
([preprint](https://www.biorxiv.org/content/early/2019/01/07/512749?rss=1))  

